File: backend/src/bond_monitor/domain/portfolio/selection.py
"""Unified bond eligibility, ranking, and profile-fallback selection."""

from __future__ import annotations

import bisect
from collections.abc import Sequence
from dataclasses import dataclass, field
from datetime import date, timedelta

from bond_monitor.domain.bonds.models import RATING_ORDER, BondRecord, RiskLevel
from bond_monitor.domain.portfolio.models import Portfolio, RiskProfile
from bond_monitor.domain.portfolio.policies import (
    DEFAULT_BOND_SELECTION_POLICY,
    DEFAULT_DURATION_POLICY,
    BondSelectionContext,
    BondSelectionPolicy,
    DurationPolicy,
)
from bond_monitor.domain.screening.scorer import score_bonds_for_profile
from bond_monitor.domain.shared.formatting import format_date

_NORMAL_MIN_RATING_ORDINAL: int = RATING_ORDER["ruA-"]
_AGGRESSIVE_MIN_RATING_ORDINAL: int = RATING_ORDER["ruBB-"]


@dataclass(frozen=True)
class BondSelectionResult:
    """Ranked eligible bonds after profile-fallback chain."""

    bonds: list[BondRecord]
    fallback_note: str
    effective_profile_filter: RiskProfile | None


@dataclass(frozen=True)
class MaturityIndex:
    """Bonds sorted by effective end date (maturity or offer) for window queries."""

    _dates: tuple[date, ...]
    _bonds: tuple[BondRecord, ...]

    @classmethod
    def build(cls, universe: Sequence[BondRecord]) -> MaturityIndex:
        dated: list[tuple[date, BondRecord]] = []
        for bond in universe:
            end = bond.maturity_date or bond.offer_date
            if end is not None:
                dated.append((end, bond))
        dated.sort(key=lambda item: item[0])
        if not dated:
            return cls((), ())
        dates, bonds = zip(*dated, strict=True)
        return cls(dates, bonds)

    def bonds_between(self, min_date: date, max_date: date) -> list[BondRecord]:
        if not self._dates:
            return []
        left = bisect.bisect_left(self._dates, min_date)
        right = bisect.bisect_right(self._dates, max_date)
        return list(self._bonds[left:right])


@dataclass
class SelectionOptions:
    """Per-plan selection accelerators (maturity index + ranked-result cache)."""

    maturity_index: MaturityIndex | None = None
    ranked_cache: dict[tuple[object, ...], BondSelectionResult] = field(default_factory=dict)


def build_maturity_index(universe: Sequence[BondRecord]) -> MaturityIndex:
    return MaturityIndex.build(universe)


def has_usable_price(bond: BondRecord) -> bool:
    """Bond is purchasable when it has a positive dirty price."""
    return bond.price_per_lot_rub is not None and bond.price_per_lot_rub > 0


def api_tradable_filter(bonds: Sequence[BondRecord]) -> list[BondRecord]:
    """Keep only bonds tradable via T-Invest API."""
    return [b for b in bonds if b.api_trade_available_flag is True]


def risk_profile_filter(
    bonds: Sequence[BondRecord],
    profile: RiskProfile,
) -> list[BondRecord]:
    """Filter universe by risk profile (rating, subordination, risk level)."""
    result: list[BondRecord] = []
    for bond in bonds:
        if bond.has_default or bond.has_technical_default:
            continue

        rating_ordinal: int | None = (
            RATING_ORDER.get(bond.credit_rating) if bond.credit_rating else None
        )

        if profile == RiskProfile.NORMAL:
            if bond.subordinated_flag:
                continue
            if bond.risk_level == RiskLevel.HIGH:
                continue
            if rating_ordinal is None:
                continue
            if rating_ordinal < _NORMAL_MIN_RATING_ORDINAL:
                continue
        elif profile == RiskProfile.AGGRESSIVE:
            if rating_ordinal is not None and rating_ordinal < _AGGRESSIVE_MIN_RATING_ORDINAL:
                continue

        result.append(bond)
    return result


def portfolio_universe_filter(
    bonds: Sequence[BondRecord],
    portfolio: Portfolio,
) -> list[BondRecord]:
    """Universe filtered by portfolio strategy (profile + optional API-only)."""
    filtered = risk_profile_filter(bonds, portfolio.risk_profile)
    if portfolio.api_trade_only:
        filtered = api_tradable_filter(filtered)
    return filtered


from bond_monitor.domain.portfolio.put_offer import put_offer_buy_blocked


def _fallback_steps(profile: RiskProfile) -> tuple[RiskProfile | None, ...]:
    if profile == RiskProfile.AGGRESSIVE:
        return (RiskProfile.AGGRESSIVE, RiskProfile.NORMAL, None)
    return (RiskProfile.NORMAL, None)


def _profiles_tried_label(profile: RiskProfile) -> str:
    if profile == RiskProfile.NORMAL:
        return f"«{profile.value}» и любую без дефолта"
    return f"«{profile.value}», «{RiskProfile.NORMAL.value}» и любую без дефолта"


def _min_maturity_date(ctx: BondSelectionContext, policy: BondSelectionPolicy) -> date:
    return ctx.purchase_date + timedelta(days=policy.min_replacement_horizon_days)


def _maturity_window(ctx: BondSelectionContext, policy: BondSelectionPolicy) -> str:
    min_date = _min_maturity_date(ctx, policy)
    return f"[{format_date(min_date)}, {format_date(ctx.horizon_date)}]"


def bond_eligibility_reason(
    bond: BondRecord,
    ctx: BondSelectionContext,
    policy: BondSelectionPolicy = DEFAULT_BOND_SELECTION_POLICY,
    *,
    check_budget: bool = True,
) -> str | None:
    """Return rejection reason, or None if bond passes structural eligibility."""
    if policy.exclude_default and (bond.has_default or bond.has_technical_default):
        return "дефолт / тех.дефолт"

    if not has_usable_price(bond):
        return "нет рыночной цены"

    clean_pct = bond.last_price
    if clean_pct is not None and clean_pct < policy.min_clean_price_pct:
        return f"чистая цена {clean_pct:.1f}% < {policy.min_clean_price_pct:.0f}% номинала"

    blocked = put_offer_buy_blocked(bond, ctx.purchase_date)
    if blocked is not None:
        return blocked

    end = bond.maturity_date or bond.offer_date
    if end is None:
        return "нет даты погашения / оферты"

    min_maturity = _min_maturity_date(ctx, policy)
    if end < min_maturity:
        return f"погашение {format_date(end)} раньше окна (не ранее {format_date(min_maturity)})"
    if end > ctx.horizon_date:
        return f"погашение {format_date(end)} позже горизонта {format_date(ctx.horizon_date)}"

    if check_budget and ctx.budget_rub is not None:
        lot_cost = bond.price_per_lot_rub or 0.0
        if lot_cost > ctx.budget_rub:
            return f"лот {lot_cost:,.0f} ₽ > бюджета {ctx.budget_rub:,.0f} ₽"

    return None


def eligible_bonds(
    universe: Sequence[BondRecord],
    ctx: BondSelectionContext,
    policy: BondSelectionPolicy = DEFAULT_BOND_SELECTION_POLICY,
    *,
    profile_step: RiskProfile | None,
    check_budget: bool = True,
    selection_options: SelectionOptions | None = None,
) -> list[BondRecord]:
    """Structural + profile filter for one fallback step."""
    if selection_options is not None and selection_options.maturity_index is not None:
        search_pool: Sequence[BondRecord] = selection_options.maturity_index.bonds_between(
            _min_maturity_date(ctx, policy),
            ctx.horizon_date,
        )
    else:
        search_pool = universe

    if profile_step is not None:
        pool = risk_profile_filter(search_pool, profile_step)
    else:
        pool = [
            b
            for b in search_pool
            if not policy.exclude_default or (not b.has_default and not b.has_technical_default)
        ]

    if ctx.api_trade_only:
        pool = api_tradable_filter(pool)

    result: list[BondRecord] = []
    for bond in pool:
        if bond_eligibility_reason(bond, ctx, policy, check_budget=check_budget) is None:
            result.append(bond)
    return result


def rank_bonds(
    bonds: Sequence[BondRecord],
    profile: RiskProfile,
    *,
    key_rate: float,
    tax_rate: float,
    duration_policy: DurationPolicy = DEFAULT_DURATION_POLICY,
) -> list[BondRecord]:
    """Rank bonds using portfolio profile scoring weights."""
    return score_bonds_for_profile(
        list(bonds),
        profile,
        key_rate=key_rate,
        tax_rate=tax_rate,
        duration_policy=duration_policy,
    )


def _fallback_note(
    ctx: BondSelectionContext,
    *,
    requested_step: RiskProfile | None,
    effective_step: RiskProfile | None,
) -> str:
    if effective_step == requested_step:
        return ""
    if effective_step == RiskProfile.NORMAL:
        return (
            f"профиль «{ctx.profile.value}» — нет кандидатов в окне "
            f"[{format_date(ctx.purchase_date)}, {format_date(ctx.horizon_date)}]; "
            f"выбрана бумага под NORMAL-профиль"
        )
    if effective_step is None:
        profiles_tried = (
            f"профиль «{ctx.profile.value}»"
            if ctx.profile == RiskProfile.NORMAL
            else f"профили «{ctx.profile.value}» и «{RiskProfile.NORMAL.value}»"
        )
        return (
            f"{profiles_tried} не дали кандидатов в окне; "
            "выбрана лучшая по скору бумага без профильных ограничений"
        )
    return ""


def explain_selection_failure(
    universe: Sequence[BondRecord],
    ctx: BondSelectionContext,
    policy: BondSelectionPolicy = DEFAULT_BOND_SELECTION_POLICY,
) -> str:
    """Human-readable reason when no bond could be selected."""
    profiles_tried = _profiles_tried_label(ctx.profile)

    if ctx.budget_rub is not None and ctx.budget_rub <= 0:
        return f"ожидаемый кэш {ctx.budget_rub:,.0f} ₽ ≤ 0"

    min_maturity_date = _min_maturity_date(ctx, policy)
    window = _maturity_window(ctx, policy)

    if min_maturity_date > ctx.horizon_date:
        return (
            f"окно реинвестиции слишком узкое — покупка с {format_date(ctx.purchase_date)}, "
            f"но мин. срок удержания {policy.min_replacement_horizon_days} дн. → "
            f"погашение замены не ранее {format_date(min_maturity_date)}, "
            f"а горизонт плана {format_date(ctx.horizon_date)}"
        )

    in_window: list[BondRecord] = []
    too_expensive: list[tuple[BondRecord, float]] = []
    budget = ctx.budget_rub

    for step in _fallback_steps(ctx.profile):
        for bond in eligible_bonds(universe, ctx, policy, profile_step=step, check_budget=False):
            if budget is not None:
                lot_cost = bond.price_per_lot_rub or 0.0
                if lot_cost > budget:
                    too_expensive.append((bond, lot_cost))
                    continue
            in_window.append(bond)

    if in_window:
        return (
            f"пробовали {profiles_tried}: в окне {window} есть "
            f"{len(in_window)} подходящих по сроку и бюджету бумаг(и), "
            f"но выбрать не удалось"
        )

    if too_expensive:
        min_lot = min(cost for _, cost in too_expensive)
        budget_label = f"{budget:,.0f}" if budget is not None else "—"
        return (
            f"пробовали {profiles_tried}: в окне {window} есть "
            f"{len(too_expensive)} бумаг(и), но мин. лот {min_lot:,.0f} ₽ "
            f"больше доступных {budget_label} ₽"
        )

    budget_suffix = (
        f"при доступных {ctx.budget_rub:,.0f} ₽"
        if ctx.budget_rub is not None
        else "по заданным критериям"
    )
    return (
        f"пробовали {profiles_tried}: в окне {window} "
        f"нет бумаг с погашением {budget_suffix} "
        f"(с учётом цены, лота и пут-оферт)"
    )


def _ranked_cache_key(
    ctx: BondSelectionContext,
    *,
    key_rate: float,
    tax_rate: float,
    duration_policy: DurationPolicy,
) -> tuple[object, ...]:
    budget = None if ctx.budget_rub is None else round(ctx.budget_rub)
    return (
        ctx.profile,
        ctx.horizon_date,
        ctx.purchase_date,
        budget,
        ctx.api_trade_only,
        key_rate,
        tax_rate,
        duration_policy,
    )


def select_ranked_bonds(
    universe: Sequence[BondRecord],
    ctx: BondSelectionContext,
    policy: BondSelectionPolicy = DEFAULT_BOND_SELECTION_POLICY,
    *,
    key_rate: float,
    tax_rate: float,
    duration_policy: DurationPolicy = DEFAULT_DURATION_POLICY,
    selection_options: SelectionOptions | None = None,
) -> BondSelectionResult:
    """Main entry: profile fallback chain, then rank by portfolio profile."""
    min_maturity_date = _min_maturity_date(ctx, policy)
    if min_maturity_date > ctx.horizon_date:
        return BondSelectionResult([], "", None)

    cache_key = _ranked_cache_key(
        ctx, key_rate=key_rate, tax_rate=tax_rate, duration_policy=duration_policy
    )
    if selection_options is not None:
        cached = selection_options.ranked_cache.get(cache_key)
        if cached is not None:
            return cached

    for step in _fallback_steps(ctx.profile):
        candidates = eligible_bonds(
            universe,
            ctx,
            policy,
            profile_step=step,
            selection_options=selection_options,
        )
        if not candidates:
            continue
        ranked = rank_bonds(
            candidates,
            ctx.profile,
            key_rate=key_rate,
            tax_rate=tax_rate,
            duration_policy=duration_policy,
        )
        if not ranked:
            continue
        note = _fallback_note(ctx, requested_step=ctx.profile, effective_step=step)
        result = BondSelectionResult(ranked, note, step)
        if selection_options is not None:
            selection_options.ranked_cache[cache_key] = result
        return result

    empty = BondSelectionResult([], "", None)
    if selection_options is not None:
        selection_options.ranked_cache[cache_key] = empty
    return empty


def select_best_bond(
    universe: Sequence[BondRecord],
    ctx: BondSelectionContext,
    policy: BondSelectionPolicy = DEFAULT_BOND_SELECTION_POLICY,
    *,
    key_rate: float,
    tax_rate: float,
    duration_policy: DurationPolicy = DEFAULT_DURATION_POLICY,
    selection_options: SelectionOptions | None = None,
) -> tuple[BondRecord | None, str]:
    """Pick top-ranked bond; second value is note or failure reason."""
    if ctx.budget_rub is not None and ctx.budget_rub <= 0:
        return None, explain_selection_failure(universe, ctx, policy)

    result = select_ranked_bonds(
        universe,
        ctx,
        policy,
        key_rate=key_rate,
        tax_rate=tax_rate,
        duration_policy=duration_policy,
        selection_options=selection_options,
    )
    if result.bonds:
        return result.bonds[0], result.fallback_note

    return None, explain_selection_failure(universe, ctx, policy)

File: backend/internal/domain/portfolio/selection.go
package portfolio

import (
	"fmt"
	"math"
	"sort"
	"time"

	"github.com/tonatos/bond-monitor/backend/internal/domain/bonds"
	"github.com/tonatos/bond-monitor/backend/internal/domain/screening"
	"github.com/tonatos/bond-monitor/backend/internal/domain/shared"
)

const (
	conservativeMinRating = 7  // ruA
	normalMinRating       = 6  // ruA-
	aggressiveMinRating   = 0  // ruBB-
)

type BondSelectionResult struct {
	Bonds                  []bonds.BondRecord
	FallbackNote           string
	EffectiveProfileFilter *RiskProfile
}

type MaturityIndex struct {
	dates []time.Time
	bonds []bonds.BondRecord
}

func BuildMaturityIndex(universe []bonds.BondRecord) MaturityIndex {
	var dated []struct {
		d time.Time
		b bonds.BondRecord
	}
	for _, bond := range universe {
		end := bond.MaturityDate
		if end == nil {
			end = bond.OfferDate
		}
		if end != nil {
			dated = append(dated, struct {
				d time.Time
				b bonds.BondRecord
			}{*end, bond})
		}
	}
	sort.Slice(dated, func(i, j int) bool { return dated[i].d.Before(dated[j].d) })
	idx := MaturityIndex{}
	for _, item := range dated {
		idx.dates = append(idx.dates, item.d)
		idx.bonds = append(idx.bonds, item.b)
	}
	return idx
}

func (m MaturityIndex) BondsBetween(minDate, maxDate time.Time) []bonds.BondRecord {
	if len(m.dates) == 0 {
		return nil
	}
	left := sort.Search(len(m.dates), func(i int) bool { return !m.dates[i].Before(minDate) })
	right := sort.Search(len(m.dates), func(i int) bool { return m.dates[i].After(maxDate) })
	return append([]bonds.BondRecord(nil), m.bonds[left:right]...)
}

type SelectionOptions struct {
	MaturityIndex *MaturityIndex
	RankedCache   map[string]BondSelectionResult
}

func HasUsablePrice(b bonds.BondRecord) bool {
	p := b.PricePerLotRub()
	return p != nil && *p > 0
}

func APITradableFilter(bs []bonds.BondRecord) []bonds.BondRecord {
	var out []bonds.BondRecord
	for _, b := range bs {
		if b.APITradeAvailableFlag != nil && *b.APITradeAvailableFlag {
			out = append(out, b)
		}
	}
	return out
}

func ratingOrdinal(b bonds.BondRecord) *int {
	if b.CreditRating == nil {
		return nil
	}
	if ord, ok := bonds.RatingOrder[*b.CreditRating]; ok {
		return &ord
	}
	return nil
}

func RiskProfileFilter(bs []bonds.BondRecord, profile RiskProfile) []bonds.BondRecord {
	var result []bonds.BondRecord
	for _, bond := range bs {
		if bond.HasDefault || bond.HasTechnicalDefault {
			continue
		}
		ord := ratingOrdinal(bond)
		switch profile {
		case RiskProfileConservative:
			if bond.SubordinatedFlag || bond.RiskLevel == bonds.RiskLevelHigh || ord == nil || *ord < conservativeMinRating || bond.CallDate != nil {
				continue
			}
		case RiskProfileNormal:
			if bond.SubordinatedFlag || bond.RiskLevel == bonds.RiskLevelHigh || ord == nil || *ord < normalMinRating {
				continue
			}
		case RiskProfileAggressive:
			if ord != nil && *ord < aggressiveMinRating {
				continue
			}
		}
		result = append(result, bond)
	}
	return result
}

func PortfolioUniverseFilter(bs []bonds.BondRecord, p Portfolio) []bonds.BondRecord {
	filtered := RiskProfileFilter(bs, p.RiskProfile)
	if p.APITradeOnly {
		filtered = APITradableFilter(filtered)
	}
	return filtered
}

func fallbackSteps(profile RiskProfile) []*RiskProfile {
	switch profile {
	case RiskProfileAggressive:
		return []*RiskProfile{profilePtr(RiskProfileAggressive), profilePtr(RiskProfileNormal), nil}
	case RiskProfileConservative:
		return []*RiskProfile{profilePtr(RiskProfileConservative), profilePtr(RiskProfileNormal), nil}
	default:
		return []*RiskProfile{profilePtr(RiskProfileNormal), nil}
	}
}

func profilePtr(p RiskProfile) *RiskProfile { return &p }

func profilesTriedLabel(profile RiskProfile) string {
	if profile == RiskProfileNormal {
		return "«normal» и любую без дефолта"
	}
	return fmt.Sprintf("«%s», «normal» и любую без дефолта", profile)
}

func minMaturityDate(ctx BondSelectionContext, policy BondSelectionPolicy) time.Time {
	return shared.AddDays(ctx.PurchaseDate, policy.MinReplacementHorizonDays)
}

func maturityWindow(ctx BondSelectionContext, policy BondSelectionPolicy) string {
	minDate := minMaturityDate(ctx, policy)
	return fmt.Sprintf("[%s, %s]", shared.FormatDate(&minDate), shared.FormatDate(&ctx.HorizonDate))
}

func BondEligibilityReason(b bonds.BondRecord, ctx BondSelectionContext, policy BondSelectionPolicy, checkBudget bool) *string {
	if policy.ExcludeDefault && (b.HasDefault || b.HasTechnicalDefault) {
		msg := "дефолт / тех.дефолт"
		return &msg
	}
	if !HasUsablePrice(b) {
		msg := "нет рыночной цены"
		return &msg
	}
	if b.LastPrice != nil && *b.LastPrice < policy.MinCleanPricePct {
		msg := fmt.Sprintf("чистая цена %.1f%% < %.0f%% номинала", *b.LastPrice, policy.MinCleanPricePct)
		return &msg
	}
	if blocked := PutOfferBuyBlocked(b, ctx.PurchaseDate); blocked != nil {
		return blocked
	}
	end := b.MaturityDate
	if end == nil {
		end = b.OfferDate
	}
	if end == nil {
		msg := "нет даты погашения / оферты"
		return &msg
	}
	minMat := minMaturityDate(ctx, policy)
	if end.Before(minMat) {
		msg := fmt.Sprintf("погашение %s раньше окна (не ранее %s)", shared.FormatDate(end), shared.FormatDate(&minMat))
		return &msg
	}
	if end.After(ctx.HorizonDate) {
		msg := fmt.Sprintf("погашение %s позже горизонта %s", shared.FormatDate(end), shared.FormatDate(&ctx.HorizonDate))
		return &msg
	}
	if checkBudget && ctx.BudgetRub != nil {
		lotCost := 0.0
		if p := b.PricePerLotRub(); p != nil {
			lotCost = *p
		}
		if lotCost > *ctx.BudgetRub {
			msg := fmt.Sprintf("лот %s ₽ > бюджета %s ₽", shared.FormatNumber(lotCost, 0), shared.FormatNumber(*ctx.BudgetRub, 0))
			return &msg
		}
	}
	return nil
}

func EligibleBonds(
	universe []bonds.BondRecord,
	ctx BondSelectionContext,
	policy BondSelectionPolicy,
	profileStep *RiskProfile,
	checkBudget bool,
	opts *SelectionOptions,
) []bonds.BondRecord {
	var searchPool []bonds.BondRecord
	if opts != nil && opts.MaturityIndex != nil {
		searchPool = opts.MaturityIndex.BondsBetween(minMaturityDate(ctx, policy), ctx.HorizonDate)
	} else {
		searchPool = universe
	}
	var pool []bonds.BondRecord
	if profileStep != nil {
		pool = RiskProfileFilter(searchPool, *profileStep)
	} else {
		for _, b := range searchPool {
			if !policy.ExcludeDefault || (!b.HasDefault && !b.HasTechnicalDefault) {
				pool = append(pool, b)
			}
		}
	}
	if ctx.APITradeOnly {
		pool = APITradableFilter(pool)
	}
	var result []bonds.BondRecord
	for _, bond := range pool {
		if BondEligibilityReason(bond, ctx, policy, checkBudget) == nil {
			result = append(result, bond)
		}
	}
	return result
}

func RankBonds(
	candidates []bonds.BondRecord,
	universe []bonds.BondRecord,
	profile RiskProfile,
	keyRate, taxRate float64,
	dp DurationPolicy,
) []bonds.BondRecord {
	if len(candidates) == 0 {
		return nil
	}
	screenDP := toScreeningDurationPolicy(dp)
	screenProfile := screening.RiskProfile(profile)
	durationScale := screening.DurationScaleYears(universe, screenDP)

	byISIN := make(map[string]bonds.BondRecord, len(universe))
	for _, b := range universe {
		byISIN[b.ISIN] = b
	}

	result := make([]bonds.BondRecord, 0, len(candidates))
	for _, candidate := range candidates {
		bond := candidate
		if scored, ok := byISIN[candidate.ISIN]; ok && len(scored.ProfileScores) > 0 {
			bond = scored
		} else {
			fallback := screening.ScoreBondsForProfile(
				[]bonds.BondRecord{candidate},
				screenProfile,
				keyRate,
				taxRate,
				screenDP,
			)
			if len(fallback) > 0 {
				bond = fallback[0]
			}
		}
		resolved := screening.ResolveProfileScores(&bond, screenDP, durationScale)
		if score, ok := resolved[string(profile)]; ok {
			bond.Score = &score
			bond.ProfileScores = resolved
		}
		result = append(result, bond)
	}
	sort.Slice(result, func(i, j int) bool {
		si, sj := 0.0, 0.0
		if result[i].Score != nil {
			si = *result[i].Score
		}
		if result[j].Score != nil {
			sj = *result[j].Score
		}
		return si > sj
	})
	return result
}

func toScreeningDurationPolicy(dp DurationPolicy) screening.DurationPolicy {
	return screening.DurationPolicy{
		RateScenario:             screening.RateScenario(dp.RateScenario),
		DurationScoreWeight:      dp.DurationScoreWeight,
		TargetDurationYears:      dp.TargetDurationYears,
		MaxWeightedDurationYears: dp.MaxWeightedDurationYears,
		FloaterRateDurationYears: dp.FloaterRateDurationYears,
	}
}

func fallbackNote(ctx BondSelectionContext, requested, effective *RiskProfile) string {
	if effective == nil || (requested != nil && *effective == *requested) {
		return ""
	}
	if *effective == RiskProfileNormal {
		return fmt.Sprintf(
			"профиль «%s» — нет кандидатов в окне [%s, %s]; выбрана бумага под NORMAL-профиль",
			ctx.Profile, shared.FormatDate(&ctx.PurchaseDate), shared.FormatDate(&ctx.HorizonDate),
		)
	}
	profiles := fmt.Sprintf("профиль «%s»", ctx.Profile)
	if ctx.Profile != RiskProfileNormal {
		profiles = profilesTriedLabel(ctx.Profile)
	}
	return profiles + " не дали кандидатов в окне; выбрана лучшая по скору бумага без профильных ограничений"
}

func ExplainSelectionFailure(universe []bonds.BondRecord, ctx BondSelectionContext, policy BondSelectionPolicy) string {
	profiles := profilesTriedLabel(ctx.Profile)
	if ctx.BudgetRub != nil && *ctx.BudgetRub <= 0 {
		return fmt.Sprintf("ожидаемый кэш %s ₽ ≤ 0", shared.FormatNumber(*ctx.BudgetRub, 0))
	}
	minMat := minMaturityDate(ctx, policy)
	window := maturityWindow(ctx, policy)
	if minMat.After(ctx.HorizonDate) {
		return fmt.Sprintf(
			"окно реинвестиции слишком узкое — покупка с %s, но мин. срок удержания %d дн. → погашение замены не ранее %s, а горизонт плана %s",
			shared.FormatDate(&ctx.PurchaseDate), policy.MinReplacementHorizonDays,
			shared.FormatDate(&minMat), shared.FormatDate(&ctx.HorizonDate),
		)
	}
	var inWindow []bonds.BondRecord
	var tooExpensive []struct {
		b    bonds.BondRecord
		cost float64
	}
	budget := ctx.BudgetRub
	steps := fallbackSteps(ctx.Profile)
	for _, profileStep := range steps {
		for _, bond := range EligibleBonds(universe, ctx, policy, profileStep, false, nil) {
			if budget != nil {
				lotCost := 0.0
				if p := bond.PricePerLotRub(); p != nil {
					lotCost = *p
				}
				if lotCost > *budget {
					tooExpensive = append(tooExpensive, struct {
						b    bonds.BondRecord
						cost float64
					}{bond, lotCost})
					continue
				}
			}
			inWindow = append(inWindow, bond)
		}
	}
	if len(inWindow) > 0 {
		return fmt.Sprintf("пробовали %s: в окне %s есть %d подходящих по сроку и бюджету бумаг(и), но выбрать не удалось", profiles, window, len(inWindow))
	}
	if len(tooExpensive) > 0 {
		minLot := tooExpensive[0].cost
		for _, te := range tooExpensive[1:] {
			minLot = math.Min(minLot, te.cost)
		}
		budgetLabel := "—"
		if budget != nil {
			budgetLabel = shared.FormatNumber(*budget, 0)
		}
		return fmt.Sprintf("пробовали %s: в окне %s есть %d бумаг(и), но мин. лот %s ₽ больше доступных %s ₽",
			profiles, window, len(tooExpensive), shared.FormatNumber(minLot, 0), budgetLabel)
	}
	suffix := "по заданным критериям"
	if budget != nil {
		suffix = fmt.Sprintf("при доступных %s ₽", shared.FormatNumber(*budget, 0))
	}
	return fmt.Sprintf("пробовали %s: в окне %s нет бумаг с погашением %s (с учётом цены, лота и пут-оферт)", profiles, window, suffix)
}

func rankedCacheKey(ctx BondSelectionContext, keyRate, taxRate float64, dp DurationPolicy) string {
	budget := "nil"
	if ctx.BudgetRub != nil {
		budget = fmt.Sprintf("%.0f", math.Round(*ctx.BudgetRub))
	}
	return fmt.Sprintf("%s|%s|%s|%s|%t|%.4f|%.4f|%s|%v",
		ctx.Profile, ctx.HorizonDate.Format("2006-01-02"), ctx.PurchaseDate.Format("2006-01-02"),
		budget, ctx.APITradeOnly, keyRate, taxRate, dp.RateScenario, dp.MaxWeightedDurationYears)
}

func SelectRankedBonds(
	universe []bonds.BondRecord,
	ctx BondSelectionContext,
	policy BondSelectionPolicy,
	keyRate, taxRate float64,
	dp DurationPolicy,
	opts *SelectionOptions,
) BondSelectionResult {
	if minMaturityDate(ctx, policy).After(ctx.HorizonDate) {
		return BondSelectionResult{}
	}
	cacheKey := rankedCacheKey(ctx, keyRate, taxRate, dp)
	if opts != nil && opts.RankedCache != nil {
		if cached, ok := opts.RankedCache[cacheKey]; ok {
			return cached
		}
	}
	for _, step := range fallbackSteps(ctx.Profile) {
		candidates := EligibleBonds(universe, ctx, policy, step, true, opts)
		if len(candidates) == 0 {
			continue
		}
		ranked := RankBonds(candidates, universe, ctx.Profile, keyRate, taxRate, dp)
		if len(ranked) == 0 {
			continue
		}
		req := ctx.Profile
		result := BondSelectionResult{Bonds: ranked, FallbackNote: fallbackNote(ctx, &req, step), EffectiveProfileFilter: step}
		if opts != nil {
			if opts.RankedCache == nil {
				opts.RankedCache = make(map[string]BondSelectionResult)
			}
			opts.RankedCache[cacheKey] = result
		}
		return result
	}
	empty := BondSelectionResult{}
	if opts != nil {
		if opts.RankedCache == nil {
			opts.RankedCache = make(map[string]BondSelectionResult)
		}
		opts.RankedCache[cacheKey] = empty
	}
	return empty
}

func SelectBestBond(
	universe []bonds.BondRecord,
	ctx BondSelectionContext,
	policy BondSelectionPolicy,
	keyRate, taxRate float64,
	dp DurationPolicy,
	opts *SelectionOptions,
) (*bonds.BondRecord, string) {
	if ctx.BudgetRub != nil && *ctx.BudgetRub <= 0 {
		return nil, ExplainSelectionFailure(universe, ctx, policy)
	}
	result := SelectRankedBonds(universe, ctx, policy, keyRate, taxRate, dp, opts)
	if len(result.Bonds) > 0 {
		return &result.Bonds[0], result.FallbackNote
	}
	return nil, ExplainSelectionFailure(universe, ctx, policy)
}
